src/phpDocumentor/Dsn.php

```diff
@@ -0,0 +1,308 @@
+     * @return string
+     */
+    public function getHost()
+    {
+        return $this->host;
+    }
+
+    /**
+     * @return int
+     */
+    public function getPort()
+    {
+        return $this->port;
+    }
+
+    /**
+     * @return string
+     */
+    public function getUsername()
+    {
+        return $this->user;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPassword()
+    {
+        return $this->password;
+    }
+
+    /**
+     * @return string
+     */
+    public function getPath()
+    {
+        return new Path($this->path);
+    }
+
+    /**
+     * @return string
+     */
+    public function getQuery()
+    {
+        return $this->query;
+    }
+
+    /**
+     * @return string
+     */
+    public function getParameters()
+    {
+        return $this->parameters;
+    }
+
+    /**
+     * @param $dsn
+     */
+    private function parse($dsn)
+    {
+        $this->dsnParts = explode(';', $dsn);
+        $location = $this->dsnParts[0];
+        unset($this->dsnParts[0]);
+        $this->locationParts = parse_url($location);
+
+        if (! $this->locationParts) {
+            throw new \InvalidArgumentException(
+                sprintf('"%s" is not a valid DSN.', $dsn)
+            );
+        }
+
+
+        $this->parseScheme();
+
+        $this->parseHostAndPath();
+
+        $this->parsePort();
+
+        $this->parseUser();
+
+        $this->parsePassword();
+
+        $this->parseQuery();
+
+        $this->parseParameters();
+    }
+
+    /**
+     * validates and sets the scheme property
+     */
+    private function parseScheme()
+    {
+        if (array_key_exists('scheme', $this->locationParts)) {
+            $validSchemes = ['file', 'git+http', 'git+https'];
+            if (! in_array(strtolower($this->locationParts['scheme']), $validSchemes)) {
+                throw new \InvalidArgumentException(
+                    sprintf('"%s" is not a valid scheme.', $this->locationParts['scheme'])
+                );
+            }
+            $this->scheme = $this->locationParts['scheme'];
+
+        } else {
+
+            $this->scheme = 'file';
+        }
+    }
+
+    /**
+     * Validates and sets the host and path properties
+     */
+    private function parseHostAndPath()
+    {
+        if (array_key_exists('scheme', $this->locationParts) && $this->locationParts['scheme'] === 'file') {
+            $path = array_key_exists('path', $this->locationParts) ? $this->locationParts['path'] : "";
+            $path = $this->locationParts['host'] . $path;
+            $result = preg_match("/^([a-zA-Z0-9-_.~\/]*(%([A-F]|[0-9]){2})*)*$/", $path);
+            if ($result === 0) {
+                throw new \InvalidArgumentException(
+                    sprintf('"%s" is not a valid path.', $this->locationParts['path'])
+                );
+            }
+
+            $this->path = $path;
+
+        } else {
+            if (array_key_exists('host', $this->locationParts)) {
```

Comment:
You duplicate the retrieval of a 'host' here (including existance check); you can remove this duplication by setting a variable $path at the beginning of this method